### DISCUSSION TITLE
Document for sync and async

### DISCUSSION THREAD

**Discussion by guesswh0** [BeanieODM/beanie#394] [Question] [answered]
Why beanie using different `Document` for sync and async?
https://beanie-odm.dev/tutorial/sync/migrate-from-async/

```python
from beanie.sync import Document

class Product(Document):
    name: str
    price: float
```

```python
from beanie import Document

class Product(Document):
    name: str
    price: float
```

For example if I provide a library with defined documents structure and this library is used by sync and async projects, then I have to duplicate the code to support both?

**Comment by roman-right** [comment, marked as answer]
Hi,

Because methods have the name but different interfaces there (sync and async respectively).

It will probably be implemented in `2.0` as a single factory class, that will produce instances with different interfaces based on configs or context. But the amount of work to do so is not even estimated yet. I hope I'll implement 2.0 by New Year, but this date can be changed.

I would suggest or implement a factory for your lib or use code duplication.

**Reply by PetrF0X** [reply]
Hi @roman-right. Is this idea still on the roadmap now that beanie project was split into beanie (async) and bunnet (sync)?

**Reply by roman-right** [reply]
Hey. I had been playing with this idea for a week and found, that the code is getting ugly and unsupportable. That's why it was split.

I was thinking about the use case where there are sync and async docs with the same fields and additional methods - I think, the solution is a mixin with all the fields and methods, that will be provided to both document types.

As for now, I don't plan to make a merger. Maybe things would change.
